### PR TITLE
refactor the encryption/vk logic

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why this change is needed

The logic of encryption with viewing keys and error handling is fiddly, error prone and repetitive.

### What changes were made as part of this PR

- created a template and builder pattern style function that contains all the relevant VK logic. 
- refactor all methods that respond to user queries. The logic is moved to dedicated files and uses the above pattern
- refactor the vk_handler to clarify responsibilites, mostly by splitting out the address verification
- create a dedicated struct for transporting the VK across the RPC

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/ten-protocol/ten-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [ ] PR checks reviewed and performed

### CODE REVIEW

File: go/common/log_events.go [hunk 1]
```diff
@@ -23,9 +22,6 @@ type LogSubscription struct {
 
 	// A subscriber-defined filter to apply to the stream of logs.
 	Filter *filters.FilterCriteria
-
-	// Handles the viewing key encryption
-	VkHandler *vkhandler.VKHandler
```
Comment (tudor-malene) [inline go/common/log_events.go:28]: this is moved to an internal datastructure to avoid serialising it

File: go/common/viewingkey/viewing_key.go [hunk 2]
```diff
@@ -208,44 +207,40 @@ func CalculateUserID(publicKeyBytes []byte) []byte {
...
 	}
 
 	pubKey, err := crypto.UnmarshalPubkey(pubKeyBytes)
 	if err != nil {
-		return false
-	}
-
-	recoveredAddr := crypto.PubkeyToAddress(*pubKey)
-
-	if !bytes.Equal(recoveredAddr.Bytes(), address.Bytes()) {
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply go/common/viewingkey/viewing_key.go:228]: I moved that check a layer above. 
The `AuthenticatedViewingKey` is responsible for telling you what account it is for.
Then in the layer above ( the `withEncryption` function`), we check whether that account matches the "logical from" . If it doesn't, the processing stops, and an error is returned

**Review by BedrockSquirrel — COMMENTED** [review]
I 100% agree with the aims of this PR (to reduce repetition and make the methods easier to read and focused on business logic) but I think how it is at the moment still puts a lot on the dev in terms of readability/risk of errors.

I like the abstraction but maybe if we could lean into go's types and stuff a bit more so it was self-documenting and harder for devs to make a mistake I think that would help.

I think it would be possible and helpful to get the `withVKEncryption` method signature looking something more like this:
```
func (e *enclaveImpl) executeVKEncryptedRequest[P RequestParams, R RequestResponse](
	encReq []byte,
	extractRequest func([]any) (Request[P], error),
	execute func(Request[P]) (R, error),
) (*responses.EnclaveResponse, common.SystemError) {
```
where Request is defined like:
```
type Request[P any] struct {
   Sender *gethcommon.Address
   Params P
}
```

So then using it, our enclave methods would look more like this:
```
func (e *enclaveImpl) ObsCall(encryptedParams common.EncryptedParamsCall) (*responses.Call, common.SystemError) {
   return e.executeVKEncryptedRequest[ObsCallReq, ObsCallResp](
      encryptedParams,

      func(reqParams []any) (*EncRequest[ObsCallReq], error) {
         // ... code extracting the fields .. //

         return &EncRequest{Sender: &sender, Params: ObsCallReq{ApiArgs: apiArgs, BlkNumber: blkNumber}}, nil
      },

      func(callReq *EncRequest[ObsCallReq]) (ObsCallResp, error) {
         // now in the execute function you don't have to do the casting

         // can access callReq.Sender
         // can access callReq.Params.ApiArgs and callReq.Params.BlkNumber (with typed params, so autocomplete etc. works)
      }
   )
}
```

This was just off top of my head so maybe there's some barrier here I've not seen and I'm sure we can do better than the above. But yeah tldr I think it might worth:
- Using generics for types so we don't need casting and comments explaining types
- going back to having a combined error value rather than having execute return two positional errors (or could bundle errors for users into `RequestResponse` if that makes more sense). Wrapping the error with its intention is easier for the dev than remembering if it's the first err return value or the second one imo.

File: go/enclave/events/subscription_manager.go [hunk 3]
```diff
@@ -90,9 +93,9 @@ func (s *SubscriptionManager) RemoveSubscription(id gethrpc.ID) {
 }
 
 // FilterLogsForReceipt removes the logs that the sender of a transaction is not allowed to view
-func (s *SubscriptionManager) FilterLogsForReceipt(receipt *types.Receipt, account *gethcommon.Address) ([]*types.Log, error) {
+func FilterLogsForReceipt(receipt *types.Receipt, account *gethcommon.Address, storage storage.Storage) ([]*types.Log, error) {
```
Comment (tudor-malene) [inline go/enclave/events/subscription_manager.go:96]: transformed this to a simple function to break a dependency cycle

File: go/enclave/rpc/EstimateGas.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+			return &UserResponse[hexutil.Uint64]{&totalGasEstimate, nil}, nil
+		})
+}
+
+// DoEstimateGas returns the estimation of minimum gas required to execute transaction
+// This is a copy of https://github.com/ethereum/go-ethereum/blob/master/internal/ethapi/api.go#L1055
+// there's a high complexity to the method due to geth business rules (which is mimic'd here)
+// once the work of obscuro gas mechanics is established this method should be simplified
+func (rpc *EncryptionManager) doEstimateGas(args *gethapi.TransactionArgs, blkNumber *gethrpc.BlockNumber, gasCap uint64) (hexutil.Uint64, common.SystemError) { //nolint: gocognit
```
Comment (BedrockSquirrel) [inline go/enclave/rpc/EstimateGas.go:111]: Really nice that all this implementation detail is in a dedicated file rather than dumped in enclave.go

File: go/enclave/rpc/GetBalance.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+			if err != nil {
+				return nil, fmt.Errorf("unable to extract requested address - %w", err)
+			}
+
+			blockNumber, err := gethencoding.ExtractBlockNumber(reqParams[1])
+			if err != nil {
+				return nil, fmt.Errorf("unable to extract requested block number - %w", err)
+			}
+
+			encryptAddress, balance, err := rpc.chain.GetBalance(*requestedAddress, blockNumber)
```
Comment (BedrockSquirrel) [inline go/enclave/rpc/GetBalance.go:33]: I think the execution has slipped into the parameters function here, and then you pass the balance result into the next part as if it was a param.

Maybe we're thinking about something wrong if the distinction between these two functions isn't that important?

**Review by BedrockSquirrel — COMMENTED** [review]
Getting all the implementation details into dedicated files feels like a step in the right direction. This stuff is all still super fiddly, couple of times I wrote comments then had to go back and delete them after a bit because I read more and got a "oh that's why it had to be like that" moment.

Feel like it's still a little bit hard to use with the anonymous functions and the 1 and 2 versions of things (we'd be in a world of pain if we ever wanted 3 args haha).

And the definition between the parameter extraction and the execution is sometimes a bit blurry, like getTransaction where we get the transaction in the parameter extraction part because we want to know the address threw me off for a bit.

But yeah maybe it's getting into subjective opinion now so happy to sign off if you think it's good enough.

File: go/enclave/rpc/GetBalance.go (hunk 5 above)
Comment (tudor-malene) [reply go/enclave/rpc/GetBalance.go]: Sometimes, in order to extract the sender, you need to execute.
And since the call is already executed, the "Execute" becomes trivial.

In a world where performance wouldn't matter, you'd execute whatever you need to get the sender in the first call. And then re-execute in the second for the actual result

File: go/enclave/rpc/vk_utils.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	"fmt"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ten-protocol/go-ten/go/common"
+	"github.com/ten-protocol/go-ten/go/enclave/vkhandler"
+	"github.com/ten-protocol/go-ten/go/responses"
+)
+
+// UserRPCRequest1 - decoded RPC argument accompanied by a logical sender
+type UserRPCRequest1[P any] struct {
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply go/enclave/rpc/vk_utils.go:14]: if you need 3, you can create a dedicated struct, and use the "1" again.

The 2 is really a convenience to avoid the need to create yet another struct.

File: go/common/log_events.go [hunk 7]
```diff
@@ -5,20 +5,13 @@ import (
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/eth/filters"
 	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ten-protocol/go-ten/go/common/viewingkey"
 )
 
 // LogSubscription is an authenticated subscription to logs.
 type LogSubscription struct {
-	// The account the events relate to.
-	Account *common.Address
```
Comment (tudor-malene) [inline go/common/log_events.go:14]: no need to send this, because it can be retrieved from the  signature over the VK.
Even if sent, It has to be checked to match the retrieved key, and creates a security risk

File: go/enclave/storage/enclavedb/batch.go [hunk 8]
```diff
@@ -401,7 +401,7 @@ func ReadReceipt(db *sql.DB, hash common.L2TxHash, config *params.ChainConfig) (
 	return receipts[0], nil
 }
 
-func ReadTransaction(db *sql.DB, txHash gethcommon.Hash) (*types.Transaction, gethcommon.Hash, uint64, uint64, error) {
+func ReadTransaction(db *sql.DB, txHash gethcommon.Hash) (*types.Transaction, common.L2BatchHash, uint64, uint64, error) {
```
Comment (tudor-malene) [inline go/enclave/storage/enclavedb/batch.go:404]: this is an unrelated change. But it confused me, so I used the type alias to clarify the meaning

**Review by BedrockSquirrel — APPROVED** [review]
LGTM - lot to take in but gave it a read through since we went through on the call and nothing jumping out to me. Much more maintainable/debuggable now imo
